
syntax = "proto3";

// the message exchange service definition.
service MessageExchange {
  rpc CreateAccount (AccountRequest) returns (AccountResponse) {}
  rpc LogIn (AccountRequest) returns (AccountResponse) {}
  rpc ListAccounts (ListAccountsRequest) returns (ListAccountsResponse) {}
  rpc LogOut (AccountRequest) returns (AccountResponse) {}
  rpc DeleteAccount (AccountRequest) returns (AccountResponse) {}
  rpc SendMessage (SendMessageRequest) returns (SendMessageResponse) {}
  rpc RequestMessages(RequestMessagesRequest) returns (RequestMessagesResponse) {}
  rpc GetNewMessages(GetNewMessagesRequest) returns (RequestMessagesResponse) {}
}


message Account {
    string name = 1;
    optional bool loggedIn = 2;
}

message Message {
    string sender_id = 1;
    string receiver_id = 2;
    string message = 3;
    string timestamp = 4;
}

/*
/ Create, LogIn, LogOut, and Delete Account
*/ 

message AccountRequest {
    string name = 1;
}

message AccountResponse {
    int32 response_code = 1;
    string response_text = 2;
}

message GetNewMessagesRequest {
    string sender_id = 1;    
}


/*
/ List Accounts
*/ 
message ListAccountsRequest {
    string search_pattern = 1;
}

message ListAccountsResponse {
    repeated Account accounts = 1;
}

/*
/ Send Message
*/ 
message SendMessageRequest {
    string sender_id = 1;
    string receiver_id = 2;
    string message = 3;
}

message SendMessageResponse {
    int32 response_code = 1;
    string delivered = 2;
}

/*
/ Request Messages
*/

message RequestMessagesRequest {
    string sender_id = 1;
    string receiver_id = 2;
}

message RequestMessagesResponse {
    int32 response_code = 1;
    repeated Message messages = 2;
    optional string error = 3;
}